yaziPlugins: init

khaneliman opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done
Initialize a yaziPlugins set with the latest compatible versions of the yazi-rs/plugins. 

Closes https://github.com/NixOS/nixpkgs/issues/360867
Follow up TODO:

- [x] Update script to check what the `@since` version in the plugin is and only update when current yazi version supports it
- [x] Include other yazi plugins, outside of main repo

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

khaneliman committed:
yaziPlugins.full-border: init at 25.2.26-unstable-2025-03-11

khaneliman committed:
yaziPlugins.chmod: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.diff: init at 25.2.7-unstable-2025-03-02

khaneliman committed:
yaziPlugins.git: init at 25.2.26-unstable-2025-03-17

khaneliman committed:
yaziPlugins.jump-to-char: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.lsar: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.mactag: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.mime-ext: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.mount: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.no-status: init at 25.2.7-unstable-2025-03-02

khaneliman committed:
yaziPlugins.smart-enter: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.smart-filter: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.toggle-pane: init at 25.2.26-unstable-2025-03-02

khaneliman committed:
yaziPlugins.vcs-files: init at 25.3.7-unstable-2025-03-07

khaneliman committed:
yaziPlugins.glow: init at 0-unstable-2025-02-22

khaneliman committed:
yaziPlugins.miller: init at 0-unstable-2025-08-28

khaneliman committed:
yaziPlugins.mediainfo: init at 25.2.7-unstable-2025-04-5

khaneliman committed:
yaziPlugins.restore: init at 25.2.7-unstable-2025-04-04

khaneliman committed:
yaziPlugins.ouch: init at 0-unstable-2025-03-29

khaneliman commented:
> I was just going to request ouch.yazi along with the other two plugins I use, you beat me to it! :D Have you looked at adding https://github.com/dedukun/relative-motions.yazi or https://github.com/Rolv-Apneseth/starship.yazi yet? Would much appreciate the addition of those two :)

Yeah, I can add them. I was gonna just add a few more that I use in my config and then merge so that others can start adding plugins themselves.

uncenter commented:
Looks like there is a bit of repetitive code for the "officially" distributed plugins from yazi-rs/plugins (`installPhase` particularly) . Could we abstract that to a helper function/builder (takes a `<name>` of a plugin -> `<name>.yazi` as pname, fetch from `yazi-rs/plugins` and do the installPhase stuff... etc)?

khaneliman commented:
> Looks like there is a bit of repetitive code for the "officially" distributed plugins from yazi-rs/plugins (`installPhase` particularly) . Could we abstract that to a helper function/builder (takes a `<name>` of a plugin -> `<name>.yazi` as pname, fetch from `yazi-rs/plugins` and do the installPhase stuff... etc)?

I was debating that, for sure... the main goal was to allow different source versions between the official repo still per a conversation on https://github.com/yazi-rs/plugins/issues/100

khaneliman committed:
yaziPlugins.sudo: init at 0-unstable-2025-02-08

khaneliman committed:
yaziPlugins.relative-motions: init at 25.2.7-unstable-2025-04-07

khaneliman committed:
yaziPlugins.starship: init at 25.2.7-unstable-2025-02-23

khaneliman committed:
yaziPlugins.bypass: init at 0-unstable-2025-02-16

khaneliman committed:
yaziPlugins: use a helper function for creating derivations

khaneliman commented:
@uncenter alright, tried adding a helper function to dedupe some of the logic

uncenter commented:
> @uncenter alright, tried adding a helper function to dedupe some of the logic

Sweet, that looks great!

khaneliman committed:
yaziPlugins.git: 25.2.26-unstable-2025-03-17 -> 25.4.4-unstable-2025-04-04

khaneliman committed:
yaziPlugins.mime-ext: 25.2.26-unstable-2025-03-17 -> 25.4.4-unstable-2025-04-04

khaneliman committed:
yaziPlugins.mactag: 25.2.26-unstable-2025-03-17 -> 25.4.4-unstable-2025-04-04

khaneliman commented:
Alright testing locally, a few packages needed to be bumped to cooperate with our latest yazi bump.

LazyStability commented:
I did start writing a bash script, which updates plugins, depending on the `@source` attribute. 

I am however stuck on update-source-version throwing an error
<details>

```bash
  #!/usr/bin/env nix-shell
  #!nix-shell -i bash -p bash nix-update curl coreutils jq common-updater-scripts gawk
  
  set -eux
  
  NIXPKGS_DIR="$PWD"
  
  # TODO: This has to be made available to the script and not hardcoded
  NAME="mount"
  PNAME="mount.yazi"
  
  OWNER=$(nix eval --raw -f "$NIXPKGS_DIR" yaziPlugins."$NAME".src.owner)
  REPO=$(nix eval --raw -f "$NIXPKGS_DIR" yaziPlugins."$NAME".src.repo)
  YAZI_VERSION=$(nix eval --raw -f "$NIXPKGS_DIR" yazi-unwrapped.version)
  
  # Get the updated plugin
  URL=https://raw.githubusercontent.com/"$OWNER"/"$REPO"/refs/heads/main
  [[ $OWNER == "yazi-rs" ]] &&  URL="$URL/$PNAME/main.lua" || URL="$URL/main.lua"
  UPDATED_PLUGIN=$(curl --silent ${GITHUB_TOKEN:+-u ":$GITHUB_TOKEN"} "$URL")
  
  # Query for the version
  RAW_REQUIRED_VERSION=$(printf "%s" "$UPDATED_PLUGIN" |head -n 1 )
  REQUIRED_VERSION=${RAW_REQUIRED_VERSION#*since\ }
  
  printf "%s\n" "$REQUIRED_VERSION" "$YAZI_VERSION"
  if  ! printf "%s\n" "$REQUIRED_VERSION" "$YAZI_VERSION" | sort -C -V; then
     echo "$NAME-plugin is unable to update"
     exit 0
  fi
  
  echo "Upating code source"
  # FIX: update-source-version: error: Couldn't locate old source hash 'sha256-nZ8yfnKvNLM5aA+mmQ3PkfM5lwSKwWnkQewcg9GwseI=' (or it appeared more than once) in '/data/Software/nixpkgs/pkgs/by-name/ya/yazi/plugins/default.nix'!
  update-source-version yaziPlugins.$NAME HEAD --source-key=src
```
<summary>update-script.sh</summary>

khaneliman commented:
> I did start writing a bash script, which updates plugins, depending on the `@source` attribute.
> 
> I am however stuck on update-source-version throwing an error
> 
> update-script.sh

Awesome, I can take a look in a little bit. Thanks for starting on that!

khaneliman committed:
yaziPlugins: add updateScript

Checks if the packaged yazi version supports the plugin's required
version before updating derivation.

Co-authored-by: LazilyStableProton <LazilyStable@proton.me>

khaneliman commented:
Alright, tested update script with a `yazi-rs` plugin and one outside repo. Also tested with and without `@since` in the file. Made script assume no `@since` directive means that it is compatible and will need manual verification.

LazyStability commented:
Running the update script with this command `nix-shell maintainers/scripts/update.nix --argstr package yaziPlugins.chmod` ([Which the docs suggest as the intended way](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#how-are-update-scripts-executed)) causes it to fail on line 28. Apparently the Github Token is not set when calling the script this way. 
The error is originating from my orginal draft, which I just ran directly. How are you executing the script?

khaneliman commented:
> ~Running the update script with this command `nix-shell maintainers/scripts/update.nix --argstr package yaziPlugins.chmod` ([Which the docs suggest as the intended way](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#how-are-update-scripts-executed)) causes it to fail on line 28. Apparently the Github Token is not set when calling the script this way. The error is originating from my orginal draft, which I just ran directly. How are you executing the script?~
> 
> Edit: It seems like the Github Token not being set is my mistake. I am baffled by this, it did work with the `gh cli` before. It works after manually setting it.

GitHub token should be optional, if it is crashing when not present, that’s a bug.

LazyStability commented:
Then the shell parameter extension in line 28 is incorrect. It would need to be `${GITHUB_TOKEN:-}` instead of `"${GITHUB_TOKEN:-u ":$GITHUB_TOKEN"}"`

khaneliman commented:
> Then the shell parameter extension in line 28 is incorrect. It would need to be `${GITHUB_TOKEN:-}` instead of `"${GITHUB_TOKEN:-u ":$GITHUB_TOKEN"}"`

Not sure when that got changed, thanks

LazyStability commented:
I just realised that it would probably be better if the `GITHUB_HEADER` declaration is moved up and the mention of GITHUB_TOKEN is just replaced by `"${GITHUB_HEADER[@]}"`. It is more consistent and does behave more robustly. 

> > Then the shell parameter extension in line 28 is incorrect. It would need to be `${GITHUB_TOKEN:-}` instead of `"${GITHUB_TOKEN:-u ":$GITHUB_TOKEN"}"`
> 
> Not sure when that got changed, thanks
> 
> EDIT: oh nvm that was in your script that I started with

I am sadly not very proficient at bash.

khaneliman committed:
yaziPlugins: rewrite update script in python

LazyStability commented:
Can you tag the yazi-plugins issue in this pr? https://github.com/NixOS/nixpkgs/issues/360867

teto reviewed (approved):
this is awesome !

khaneliman commented:
Can probably simplify the updater experience but, would like to at least get the package set created and available.
